snuba/web/db_query.py

```diff
@@ -55,6 +57,152 @@ def update_query_metadata_and_stats(
+        query_settings["max_threads"] = 1
+
+    result = reader.execute(
+        query,
+        query_settings,
+        query_id=query_id,
+        with_totals=request.query.has_totals(),
+    )
+
+    timer.mark("execute")
+    stats.update(
+        {"result_rows": len(result["data"]), "result_cols": len(result["meta"])}
+    )
+
+    return result
+
+
+def execute_query_with_rate_limits(
+    request: Request,
+    query: ClickhouseQuery,
+    timer: Timer,
+    stats: MutableMapping[str, Any],
+    query_settings: MutableMapping[str, Any],
+    query_id: Optional[str],
+) -> Result:
+    # XXX: We should consider moving this that it applies to the logical query,
+    # not the physical query.
+    with RateLimitAggregator(
+        request.settings.get_rate_limit_params()
+    ) as rate_limit_stats_container:
+        stats.update(rate_limit_stats_container.to_dict())
+        timer.mark("rate_limit")
+
+        project_rate_limit_stats = rate_limit_stats_container.get_stats(
+            PROJECT_RATE_LIMIT_NAME
+        )
+
+        if (
+            "max_threads" in query_settings
+            and project_rate_limit_stats is not None
+            and project_rate_limit_stats.concurrent > 1
+        ):
+            maxt = query_settings["max_threads"]
+            query_settings["max_threads"] = max(
+                1, maxt - project_rate_limit_stats.concurrent + 1
+            )
+
+        return execute_query(request, query, timer, stats, query_settings, query_id)
+
+
+def execute_query_with_caching(
+    request: Request,
+    query: ClickhouseQuery,
+    timer: Timer,
+    stats: MutableMapping[str, Any],
+    query_settings: MutableMapping[str, Any],
+    query_id: Optional[str],
+) -> Result:
+    # XXX: ``uncompressed_cache_max_cols`` is used to control both the result
+    # cache, as well as the uncompressed cache. These should be independent.
+    use_cache, uc_max = state.get_configs(
+        [("use_cache", settings.USE_RESULT_CACHE), ("uncompressed_cache_max_cols", 5)]
+    )
+
+    if len(request.query.get_all_referenced_columns()) > uc_max:
+        use_cache = False
+
+    stats["use_cache"] = use_cache
+
+    execute = partial(
+        execute_query_with_rate_limits,
+        request,
+        query,
+        timer,
+        stats,
+        query_settings,
+        query_id,
+    )
+
+    if use_cache:
+        key = md5(force_bytes(query.format_sql())).hexdigest()
+        result = cache.get(key)
+        timer.mark("cache_get")
+        stats["cache_hit"] = result is not None
+        if result is not None:
+            return result
+        result = execute()
+        cache.set(key, result)
+        timer.mark("cache_set")
+        return result
+    else:
+        # XXX: These timings and debug information are reported even if
+        # cache is not in use.
+        timer.mark("cache_get")
+        stats["cache_hit"] = False
+        return execute()
+
+
+def execute_query_with_deduplication(
```

Comment:
This will likely be replaced by #882, but will need to be able to exist alongside it for a while if we want to have a runtime switch for the new implementation.